File: linux/host/src/protocol.rs
use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u32 = 1;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    ClientHello {
        protocol: u32,
        token: String,
        device: String,
        max_width: u32,
        max_height: u32,
        max_fps: u32,
    },
    Answer {
        sdp: String,
    },
    IceCandidate {
        candidate: String,
        sdp_mid: Option<String>,
        sdp_mline_index: u32,
    },
    Ping {
        nonce: u64,
    },
    Pong {
        nonce: u64,
    },
    ClientStats {
        rendered_fps: f64,
        #[serde(default)]
        frames_decoded: Option<u64>,
        frames_dropped: u64,
        #[serde(default)]
        bitrate_kbps: Option<f64>,
        #[serde(default)]
        jitter_ms: Option<f64>,
        #[serde(default)]
        rtt_ms: Option<f64>,
        #[serde(default)]
        packets_received: Option<u64>,
        #[serde(default)]
        packets_lost: Option<i64>,
        decoder: Option<String>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    HelloAck {
        protocol: u32,
    },
    Offer {
        sdp: String,
    },
    IceCandidate {
        candidate: String,
        sdp_mid: Option<String>,
        sdp_mline_index: u32,
    },
    StreamConfig {
        source_width: u32,
        source_height: u32,
        encoded_width: u32,
        encoded_height: u32,
        fps: u32,
        bitrate_kbps: u32,
    },
    Ping {
        nonce: u64,
    },
    Pong {
        nonce: u64,
    },
    Error {
        code: String,
        message: String,
    },
}

impl ServerMessage {
    pub fn error(code: impl Into<String>, message: impl Into<String>) -> Self {
        Self::Error {
            code: code.into(),
            message: message.into(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn protocol_round_trip() {
        let message = ClientMessage::ClientHello {
            protocol: PROTOCOL_VERSION,
            token: "secret".into(),
            device: "SM-X400".into(),
            max_width: 2112,
            max_height: 1320,
            max_fps: 60,
        };
        let encoded = serde_json::to_string(&message).unwrap();
        assert_eq!(
            serde_json::from_str::<ClientMessage>(&encoded).unwrap(),
            message
        );
    }

    #[test]
    fn rejects_unknown_message() {
        assert!(serde_json::from_str::<ClientMessage>(r#"{"type":"telepathy"}"#).is_err());
    }

    #[test]
    fn accepts_legacy_client_stats() {
        let message =
            r#"{"type":"client_stats","rendered_fps":29.5,"frames_dropped":0,"decoder":null}"#;
        assert!(serde_json::from_str::<ClientMessage>(message).is_ok());
    }

    #[test]
    fn shared_fixtures_match_rust_messages() {
        let hello = include_str!("../../../protocol/fixtures/client_hello.json");
        let stats = include_str!("../../../protocol/fixtures/client_stats.json");
        let config = include_str!("../../../protocol/fixtures/stream_config.json");
        let error = include_str!("../../../protocol/fixtures/error.json");
        assert!(matches!(
            serde_json::from_str::<ClientMessage>(hello).unwrap(),
            ClientMessage::ClientHello { .. }
        ));
        assert!(matches!(
            serde_json::from_str::<ClientMessage>(stats).unwrap(),
            ClientMessage::ClientStats { .. }
        ));
        assert!(matches!(
            serde_json::from_str::<ServerMessage>(config).unwrap(),
            ServerMessage::StreamConfig { .. }
        ));
        assert!(matches!(
            serde_json::from_str::<ServerMessage>(error).unwrap(),
            ServerMessage::Error { .. }
        ));
    }
}

File: android/app/src/main/java/io/github/lamggm/auxscreen/Protocol.kt
package io.github.lamggm.auxscreen

import org.json.JSONObject

internal const val PROTOCOL_VERSION = 1

internal sealed interface HostMessage {
    data class HelloAck(val protocol: Int) : HostMessage
    data class Offer(val sdp: String) : HostMessage
    data class IceCandidate(
        val candidate: String,
        val sdpMid: String?,
        val sdpMLineIndex: Int,
    ) : HostMessage
    data class StreamConfig(
        val sourceWidth: Int,
        val sourceHeight: Int,
        val encodedWidth: Int,
        val encodedHeight: Int,
        val fps: Int,
        val bitrateKbps: Int,
    ) : HostMessage
    data class Ping(val nonce: Long) : HostMessage
    data class Pong(val nonce: Long) : HostMessage
    data class Error(val code: String, val message: String) : HostMessage
}

internal object SignalingProtocol {
    fun clientHello(token: String, model: String, width: Int, height: Int, maxFps: Int): String =
        JSONObject()
            .put("type", "client_hello")
            .put("protocol", PROTOCOL_VERSION)
            .put("token", token)
            .put("device", model)
            .put("max_width", width)
            .put("max_height", height)
            .put("max_fps", maxFps)
            .toString()

    fun answer(sdp: String): String = JSONObject()
        .put("type", "answer")
        .put("sdp", sdp)
        .toString()

    fun iceCandidate(candidate: String, sdpMid: String?, sdpMLineIndex: Int): String =
        JSONObject()
            .put("type", "ice_candidate")
            .put("candidate", candidate)
            .put("sdp_mid", sdpMid ?: JSONObject.NULL)
            .put("sdp_mline_index", sdpMLineIndex)
            .toString()

    fun pong(nonce: Long): String = JSONObject()
        .put("type", "pong")
        .put("nonce", nonce)
        .toString()

    fun clientStats(
        renderedFps: Double,
        framesDecoded: Long?,
        framesDropped: Long,
        bitrateKbps: Double?,
        jitterMs: Double?,
        rttMs: Double?,
        packetsReceived: Long?,
        packetsLost: Long?,
        jitterBufferDelayMs: Double?,
        jitterBufferTargetDelayMs: Double?,
        jitterBufferMinimumDelayMs: Double?,
        decodeTimeMs: Double?,
        processingDelayMs: Double?,
        decoder: String?,
    ): String = JSONObject()
        .put("type", "client_stats")
        .put("rendered_fps", renderedFps)
        .put("frames_decoded", framesDecoded ?: JSONObject.NULL)
        .put("frames_dropped", framesDropped)
        .put("bitrate_kbps", bitrateKbps ?: JSONObject.NULL)
        .put("jitter_ms", jitterMs ?: JSONObject.NULL)
        .put("rtt_ms", rttMs ?: JSONObject.NULL)
        .put("packets_received", packetsReceived ?: JSONObject.NULL)
        .put("packets_lost", packetsLost ?: JSONObject.NULL)
        .put("jitter_buffer_delay_ms", jitterBufferDelayMs ?: JSONObject.NULL)
        .put("jitter_buffer_target_delay_ms", jitterBufferTargetDelayMs ?: JSONObject.NULL)
        .put("jitter_buffer_minimum_delay_ms", jitterBufferMinimumDelayMs ?: JSONObject.NULL)
        .put("decode_time_ms", decodeTimeMs ?: JSONObject.NULL)
        .put("processing_delay_ms", processingDelayMs ?: JSONObject.NULL)
        .put("decoder", decoder ?: JSONObject.NULL)
        .toString()

    fun parse(payload: String): HostMessage {
        val json = JSONObject(payload)
        return when (val type = json.getString("type")) {
            "hello_ack" -> HostMessage.HelloAck(json.getInt("protocol"))
            "offer" -> HostMessage.Offer(json.getString("sdp"))
            "ice_candidate" -> HostMessage.IceCandidate(
                candidate = json.getString("candidate"),
                sdpMid = json.optString("sdp_mid").takeIf { it.isNotBlank() && it != "null" },
                sdpMLineIndex = json.getInt("sdp_mline_index"),
            )
            "stream_config" -> HostMessage.StreamConfig(
                sourceWidth = json.getInt("source_width"),
                sourceHeight = json.getInt("source_height"),
                encodedWidth = json.getInt("encoded_width"),
                encodedHeight = json.getInt("encoded_height"),
                fps = json.getInt("fps"),
                bitrateKbps = json.getInt("bitrate_kbps"),
            )
            "ping" -> HostMessage.Ping(json.getLong("nonce"))
            "pong" -> HostMessage.Pong(json.getLong("nonce"))
            "error" -> HostMessage.Error(json.getString("code"), json.getString("message"))
            else -> error("Unknown signaling message: $type")
        }
    }
}
